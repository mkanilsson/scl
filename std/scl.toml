[package]
name = "std"
description = "Scl's standard library"
type = "lib"
