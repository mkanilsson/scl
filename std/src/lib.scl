extern proc exit(i32) !
