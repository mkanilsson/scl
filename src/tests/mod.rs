use std::{env, fs, path::PathBuf, process::Command};

use insta::glob;
use miette::{GraphicalReportHandler, GraphicalTheme};
use nanoid::nanoid;

use crate::{codegen::Codegen, package::Package, typechecker::Checker};

#[test]
fn runtime() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    fn generate_tempfile(extension: &str) -> PathBuf {
        let name = format!("{}.{extension}", nanoid!());
        env::temp_dir().join(name)
    }

    glob!("sources/runtime/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let checked_main_package = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .unwrap();

        insta::assert_debug_snapshot!(checked_main_package);

        let all_units = checked_std_package
            .units
            .into_iter()
            .chain(checked_main_package.units.into_iter())
            .collect::<Vec<_>>();

        let mut codegener = Codegen::new(all_units, checker);
        let code = codegener.generate();

        insta::assert_snapshot!(code);

        let tmp = generate_tempfile("ssa");
        let asm = generate_tempfile("S");
        let executable = generate_tempfile("out");

        std::fs::write(&tmp, code).unwrap();

        let qbe_cmd = Command::new("qbe")
            .arg("-o")
            .arg(&asm)
            .arg(tmp)
            .output()
            .unwrap();

        if !qbe_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&qbe_cmd.stdout);
            let stderr = String::from_utf8_lossy(&qbe_cmd.stderr);

            panic!("QBE failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let gcc_cmd = Command::new("gcc")
            .arg("-o")
            .arg(&executable)
            .arg(&asm)
            .output()
            .unwrap();

        if !gcc_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&gcc_cmd.stdout);
            let stderr = String::from_utf8_lossy(&gcc_cmd.stderr);

            panic!("gcc failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let run_cmd = Command::new(executable).output().unwrap();

        let stdout = String::from_utf8_lossy(&run_cmd.stdout);
        let stderr = String::from_utf8_lossy(&run_cmd.stderr);

        insta::assert_snapshot!(stdout);
        insta::assert_snapshot!(stderr);
        insta::assert_snapshot!(run_cmd.status);
    });
}

#[test]
fn typecheck_errors() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    glob!("sources/typecheck_error/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let err = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .err()
            .unwrap();

        let mut buf = String::new();
        GraphicalReportHandler::new_themed(GraphicalTheme::none())
            .render_report(&mut buf, miette::Report::new(err).as_ref())
            .unwrap();

        insta::assert_snapshot!(buf);
    });
}

#[test]
fn parse_expr() {
    glob!("sources/parse_expr/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_expr(&source);

        insta::assert_debug_snapshot!(expr);
    });
}

#[test]
fn parse_stmt() {
    glob!("sources/parse_stmt/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_stmt(&source);

        insta::assert_debug_snapshot!(expr);
    });
}
