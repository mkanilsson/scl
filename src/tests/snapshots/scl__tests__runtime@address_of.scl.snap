---
source: src/tests/mod.rs
expression: checked_main_package
input_file: src/tests/sources/runtime/address_of.scl
---
CheckedPackage {
    package_id: ModuleId(
        3,
    ),
    units: [
        CheckedTranslationUnit {
            procs: [
                CheckedProc {
                    name: "main",
                    type_id: TypeId(
                        9,
                    ),
                    body: CheckedBlock {
                        stmts: [
                            VariableDeclaration {
                                stack_slot: StackSlotId(
                                    0,
                                ),
                                value: CheckedExpr {
                                    type_id: TypeId(
                                        2,
                                    ),
                                    lvalue: false,
                                    kind: Number(
                                        10,
                                    ),
                                },
                            },
                            VariableDeclaration {
                                stack_slot: StackSlotId(
                                    1,
                                ),
                                value: CheckedExpr {
                                    type_id: TypeId(
                                        10,
                                    ),
                                    lvalue: false,
                                    kind: AddressOfLValue {
                                        stack_slot: StackSlotId(
                                            0,
                                        ),
                                        offset: 0,
                                    },
                                },
                            },
                            Expr(
                                CheckedExpr {
                                    type_id: TypeId(
                                        3,
                                    ),
                                    lvalue: true,
                                    kind: DirectCall {
                                        name: "printf",
                                        params: [
                                            CheckedExpr {
                                                type_id: TypeId(
                                                    4,
                                                ),
                                                lvalue: false,
                                                kind: String(
                                                    "%d\\n",
                                                ),
                                            },
                                            CheckedExpr {
                                                type_id: TypeId(
                                                    1,
                                                ),
                                                lvalue: false,
                                                kind: BinOp {
                                                    lhs: CheckedExpr {
                                                        type_id: TypeId(
                                                            10,
                                                        ),
                                                        lvalue: false,
                                                        kind: AddressOfLValue {
                                                            stack_slot: StackSlotId(
                                                                0,
                                                            ),
                                                            offset: 0,
                                                        },
                                                    },
                                                    op: Equal,
                                                    rhs: CheckedExpr {
                                                        type_id: TypeId(
                                                            10,
                                                        ),
                                                        lvalue: true,
                                                        kind: StackValue(
                                                            StackSlotId(
                                                                1,
                                                            ),
                                                        ),
                                                    },
                                                },
                                            },
                                        ],
                                        variadic_after: Some(
                                            1,
                                        ),
                                        stack_slot: StackSlotId(
                                            2,
                                        ),
                                    },
                                },
                            ),
                            Return {
                                value: None,
                            },
                        ],
                        last: None,
                        type_id: TypeId(
                            0,
                        ),
                    },
                    params: [],
                    return_type: TypeId(
                        0,
                    ),
                    stack_slots: StackSlots {
                        slots: [
                            TypeId(
                                2,
                            ),
                            TypeId(
                                10,
                            ),
                            TypeId(
                                3,
                            ),
                        ],
                    },
                },
            ],
        },
    ],
}
